src/Optical/Emitters/Directions.jl

```diff
@@ -0,0 +1,206 @@
+module Directions
+
+using ....OpticSim, ...Geometry
+using ...Emitters
+using LinearAlgebra
+
+abstract type AbstractDirectionDistribution{T<:Real} end
+
+#---------------------------------------
+# Direction Distrubution Common Utilities
+#---------------------------------------
+
+export generate
+
+Base.iterate(a::AbstractDirectionDistribution, state = 1) = state > length(a) ? nothing : (generate(a, state - 1), state + 1)
+Base.getindex(a::AbstractDirectionDistribution, index) = generate(a, index)
+Base.firstindex(a::AbstractDirectionDistribution) = 0
+Base.lastindex(a::AbstractDirectionDistribution) = length(a) - 1
+Base.copy(a::AbstractDirectionDistribution) = a # most don't have any heap allocated stuff so don't really need copying
+
+
+#---------------------------------------
+# Constant Ray Direction
+#---------------------------------------
+"""
+Encapsulates a single ray direction, where the default direction is unitZ3 [0, 0, 1].
+
+```julia
+Constant(direction::Vec3{T}) where {T<:Real}
+Constant(::Type{T} = Float64) where {T<:Real}
+```
+"""
+struct Constant{T} <: AbstractDirectionDistribution{T}
+    direction::Vec3{T}
+
+    function Constant(direction::Vec3{T}) where {T<:Real}
+        return new{T}(direction)
+    end
+    
+    function Constant(::Type{T} = Float64) where {T<:Real}
+        direction = unitZ3(T)
+        return new{T}(direction)
+    end
+end
+export Constant
+
+Base.length(d::Constant{T}) where {T} = 1
```

Comment:
this use of length is not consistent with general Julia practice, which uses length to indicate the number of elements in a collection. A better name would be norm which is the pnorm of a vector:

Base.norm(d::Constant{T},pnorm=2) = norm(d.direction,pnorm)

Reply:
@BrianGun feel free to push small changes like this directly to the branch. Comments are also fine - it'll just take me a bit longer to get round to it.

Reply:
Just had a proper look at the code and it seems like Base.length is being used appropriately here.